### PR TITLE
[Merged by Bors] - feat(Wiedijk100Theorems): roots of a quartic

### PR DESCRIPTION
**Pull request by hanwenzhu**
* Added `quartic_eq_zero_iff` which gives the roots of the quartic equation in terms of a root `u` to a particular cubic equation, and a degenerate case `quartic_eq_zero_iff_of_q_eq_zero`
* Simplified the cubic proof

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

- [x] depends on: #18299

### CODE REVIEW

**Comment by Ruben-VandeVelde** [comment]
Meta-point: it's unfortunate that git doesn't recognize that you're moving the cubic. Would you mind creating a separate PR with just the changes to the cubic?

**Comment by hanwenzhu** [comment]
@Ruben-VandeVelde It's probably also because I changed the file too much. By adding the quartic proof, the file is changed more than 50% so git can't recognize it. I also tested locally if I commit changes to the cubic, and then move to a new file with added quartic, git still can't recognize the name change. Perhaps I can do a separate PR that only contains a file rename to SolutionOfCubicQuartic.lean and changes to the cubic, and in this PR add the quartic proof?

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 1]
```diff
@@ -7,39 +7,65 @@ import Mathlib.Tactic.LinearCombination
...
+    in terms of some root of the *cubic resolvent* $u^3 - pu^2 - 4ru + 4pr - q^2 = 0$
+    (`quartic_depressed_eq_zero_iff`).
+
 ## References
 
 The cubic formula was originally ported from Isabelle/HOL. The
 [original file](https://isabelle.in.tum.de/dist/library/HOL/HOL-ex/Cubic_Quartic.html) was written by Amine Chaieb.
 
+The proof of the quartic formula is similar in structure to the cubic, and uses the formulation in
+CRC Standard Mathematical Tables and Formulae, Daniel Zwillinger.
```
Comment (jcommelin) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:64]: Could you please add an entry for this in `docs/references.bib` and refer to it here?

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 2]
```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
...
+        (x ^ 2 * w ^ 2 + 8 * x ^ 2 * u + 8 * x ^ 2 * q / s - u ^ 2 + 4 * r) * hs
+    _ ↔ _ := by
+      have hv' : discrim 2 (2 * s) (u - q / s) = v * v := by rw [discrim]; linear_combination -hv
+      have hw' : discrim 2 (-(2 * s)) (u + q / s) = w * w := by rw [discrim]; linear_combination -hw
+      rw [mul_eq_zero, quadratic_eq_zero_iff hi2 hv', quadratic_eq_zero_iff hi2 hw']
+      simp [(by norm_num : (2 : K) * 2 = 4), or_assoc, or_comm]
+
+/-- **The Solution of Quartic**.
+  The roots of a quartic polynomial when q is nonzero.
+  See Beyer W. H., CRC Standard Mathematical Tables and Formulae.
```
Comment (jcommelin) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:196]: Idem dito.

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 3]
```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
...
+      linear_combination -hu + (-x ^ 2 * s ^ 2 - x ^ 2 * p + x ^ 2 * u) * hw +
+        (x ^ 2 * w ^ 2 + 8 * x ^ 2 * u + 8 * x ^ 2 * q / s - u ^ 2 + 4 * r) * hs
+    _ ↔ _ := by
+      have hv' : discrim 2 (2 * s) (u - q / s) = v * v := by rw [discrim]; linear_combination -hv
+      have hw' : discrim 2 (-(2 * s)) (u + q / s) = w * w := by rw [discrim]; linear_combination -hw
+      rw [mul_eq_zero, quadratic_eq_zero_iff hi2 hv', quadratic_eq_zero_iff hi2 hw']
+      simp [(by norm_num : (2 : K) * 2 = 4), or_assoc, or_comm]
+
+/-- **The Solution of Quartic**.
+  The roots of a quartic polynomial when q is nonzero.
```
Comment (jcommelin) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:195]:
```suggestion
  The roots of a quartic polynomial when `q` is nonzero.
```

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 4]
```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
...
+  See Beyer W. H., CRC Standard Mathematical Tables and Formulae.
+  Here, u needs to satisfy the cubic resolvent. An explicit expression of u is possible using
```
Comment (jcommelin) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:197]:
```suggestion
  Here, `u` needs to satisfy the cubic resolvent. An explicit expression of `u` is possible using
```

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 5]
```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
...
+  have h256 : (256 : K) = 2 ^ 8 := by norm_num
+  have h₁ : a * x ^ 4 + b * x ^ 3 + c * x ^ 2 + d * x + e =
+      a * (y ^ 4 + p * y ^ 2 + q * y + r) := by
+    rw [hp, hq, hr]
+    simp [field_simps, y, ha, h4, h8, h16, h256]; ring
+  have h₂ : ∀ x, a * x = 0 ↔ x = 0 := by intro x; simp [ha]
+  rw [h₁, h₂, quartic_depressed_eq_zero_iff hq_nonzero hu hs hv hw]
+  simp_rw [eq_sub_iff_add_eq]
+
+/-- The roots of a quartic polynomial when q equals zero. -/
```
Comment (jcommelin) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:224]:
```suggestion
/-- The roots of a quartic polynomial when `q` equals zero. -/
```

**Review by jcommelin — COMMENTED** [review]
Some minor comments.

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 6]
```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
...
+  let y := x + b / (4 * a)
+  have h4 : (4 : K) = 2 ^ 2 := by norm_num
+  have h8 : (8 : K) = 2 ^ 3 := by norm_num
+  have h16 : (16 : K) = 2 ^ 4 := by norm_num
+  have h256 : (256 : K) = 2 ^ 8 := by norm_num
+  have h₁ : a * x ^ 4 + b * x ^ 3 + c * x ^ 2 + d * x + e = a * (y ^ 4 + p * y ^ 2 + r) := by
+    rw [hp, hr]
+    simp [field_simps, y, ha, h4, h8, h16, h256]
+    linear_combination (1048576 * a ^ 10 * x + 262144 * a ^ 9 * b) * hqz
+  have h₂ : ∀ x, a * x = 0 ↔ x = 0 := by intro x; simp [ha]
```
Comment (Ruben-VandeVelde) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:244]: This is a lemma that you should be able to find and use directly, I think; also above

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 7]
```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
...
+  rw [h₁, h₂]
+  calc
+    _ ↔ 1 * (y ^ 2 * y ^ 2) + p * y ^ 2 + r = 0 := by
+      refine Eq.congr ?_ rfl
```
Comment (Ruben-VandeVelde) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:248]: Or perhaps
```suggestion
      refine Eq.congr_left ?_
```
?

File: Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean [hunk 8]
```diff
@@ -131,6 +157,104 @@ theorem cubic_eq_zero_iff_of_p_eq_zero (ha : a ≠ 0) (hω : IsPrimitiveRoot ω
...
+      ring
```
Comment (Ruben-VandeVelde) [inline Archive/Wiedijk100Theorems/SolutionOfCubicQuartic.lean:249]: Or if you want to shorten (not required):
```suggestion
    _ ↔ 1 * (y ^ 2 * y ^ 2) + p * y ^ 2 + r = 0 := Eq.congr_left (by ring)
```

**Review by Ruben-VandeVelde — APPROVED** [review]
Thanks! The statements are not particularly readable, but I think that's the inherent complexity of the formula.

**Comment by hanwenzhu** [comment]
Hi all, thank you for the comments. The suggested changes are implemented.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**